students/luke/s05/mailroom.py

```diff
@@ -0,0 +1,143 @@
+#!/usr/bin/env python3
+
+"""
+https://uwpce-pythoncert.github.io/PythonCertDevel/exercises/mailroom.html#exercise-mailroom
+"""
+
+# Store donors/history in dictionary
+donor_list = {"alice": [5],
+              "bob": [1, 2],
+              "carol": [1, 5, 100],
+              "dan": [5],
+              "erin": [2, 1]}
+
+# Format of thank-you notes
+THANK_YOU = "Thank you, {}, for your donation(s) of ${}!\n"
+
+# Destination file
+OUTFILE = "{}.txt"
+
+
+def driver():
+    """ Main loop for mailroom """
+    done = False
+    while not done:
+        try:
+            action = input(("Select an action and press return:\n"
+                            "[S]end a thank-you (add users and donations)\n"
+                            "[W]rite thank-you letters to file\n"
+                            "[C]reate a report\n"
+                            "[Q]uit\n"
+                            "> ")).lower()
+            while action not in "swcq":
+                action = input("Bad option, reenter > ").lower()
+
+            if action == "s":
+                add_donations()
+            elif action == "w":
+                write_thankyous()
+            elif action == "c":
+                create_report()
+            else:
+                done = True
+        except EOFError:
+            done = True
+        except KeyboardInterrupt:
+            done = True
+    return
+
+
+def build_thankyou(name):
+    """ Apply thankyou format string to name """
+    return (THANK_YOU.format(name, sum(donor_list[name])))
+
+
+def list_donors():
+    """ return list donor names """
+    donors = ""
+    for n in donor_list:
+        donors += n + "\n"
+    return donors[:-1]  # Trim trailing newline
+
+
+def add_donation(name, amount):
+    """ add donation for name.  add name if it doesn't exist """
+    if not donor_list.get(name):
+        print("Adding new donor: " + name)
+        donor_list[name] = []
+
+    donor_list[name].append(donation_amount)
```

Comment:
Check out setdefault method for dictionaries! https://www.saltycrane.com/blog/2010/02/python-setdefault-example/